### ISSUE TITLE
Proposal/Discussion:  Negation of if/while Condition

### ISSUE THREAD

**Issue by HaloFour** [dotnet/csharplang#568]
There have been a couple of proposals exploring how improve the readability of conditional operations that involve negation.  Some developers find the nesting of parenthesis to be unattractive at best and easy to miss at worst.

I would like to explore the potential of allowing for a negation prefix operator `!` to appear outside of the required parenthesis for `if` and `while` conditions so that the condition is effectively reversed:

```cs
if !(foo is Bar bar)
{
    // stuff
}

while !(finished)
{
    // stuff
}

do
{
    // stuff
}
while !(finished);
```

Certainly minor and the argument can definitely be made that the negation is even less obvious.  It also doesn't address concerns with operators that would still require parenthesis, e.g. `is`.

**Comment by lachbaer** [comment]
I give it 👍 because it believe that it improves reading and thinking much.

The downside is that it can be easily overlooked compared to `if (!( ... ))`. Also the position of the `!` should be fixed, otherwise there are 3 or more possibilities to write the same expression, leading to confusion and mistakes.
```cs
if! (foo is Bar bar) ...
if ! (foo is Bar bar) ...
if !(foo is Bar bar) ...
// worst:
if                          !
  ( foo is Bar bar ) ...
```
**foobar!** 😁 

Of all I prefer the @HaloFour style. (Hey, good name, "the HaloFour condition syntax" 😇 )

**Comment by jnm2** [comment]
I prefer the readability of `while (!isFinished)` which is already available, and if we had to choose between them, I'd rather have `if (foo isnot Bar bar)` than `if !(foo is Bar bar)`.

**Comment by lachbaer** [comment]
A little side note: though they are kind of obsolete for this case, an a priori negated condition could call the `operator false`. `if (! TypeWithOperatorTrue) ... ` gives a compilation error, unless that type has an implicit convertion to boolean as well.

**Comment by bondsbw** [comment]
> Also the position of the `!` should be fixed

I disagree, that would prevent no spaces:
```c#
if!(foo is Bar bar)
```
which is less common than `if (...` but it does exist.

**Comment by bondsbw** [comment]
@jnm2 agreed with `isnot`, or perhaps `!is` or `is!`.  Though it could coexist with this proposal (which can be used for situations other than `is`).

Both situations would need to consider how pattern variables should work.

**Comment by casperOne** [comment]
I'm pretty meh on this.  I don't think the readability is any better with the removal of the outer parenthesis (I see it that way, not moving the not operator out).

And if the readability isn't improved, ergo, there's no good reason to do it (IMO).

**Comment by CyrusNajmabadi** [comment]
Rather than this, i would just prefer that we relax the grammar for ```if/while/switch/etc.``` to not be ```if ( expr )``` but just have it be ```if expr```.

if we did that, then ```if ! ...``` just calls out.    

Currently the requirement of parens is simply a holdover from C.  There is no actual syntactic need/ambiguity it addresses.  Which construct this is is already governed by the leading keyword.   

--

Another way to think about this is the grammar was always ```if expr```, but there was normative rules that state that ```expr``` needed to be a parenthesized-expression.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Indeed, that is another option.  I probably wouldn't prefer that syntax in the normal cases, but it does enable the syntax above as well.  Personally I like the parenthesis and I apply them generally anywhere that I might have a non-trivial expression, such as with `return` statements or ternary conditions.

**Comment by DavidArno** [comment]
If you championed that, @CyrusNajmabadi, I'd be more than happy to promise to never ever again argue with you or make snide comments about the motives and behaviour of the team.

Deal? 😀

**Comment by HaloFour** [comment]
#itsatrap

**Comment by CyrusNajmabadi** [comment]
I would do it... if i believed you :D

**Comment by casperOne** [comment]
> Rather than this, i would just prefer that we relax the grammar for `if/while/switch/etc.` to not be `if ( expr )` but just have it be `if expr`.

Does this mean that one liners such as this:

````
var cond = true;

if cond Console.WriteLine("true");
````

Would be legal?

When separating on multiple lines, the syntax looks OK, but when on the same line, it starts to get all jumbled together.

At least that's my perception.  This becomes a stylistic issue though, I don't see any real drawback to having this.

Note, I don't see any benefit, but like I said, it's stylistic, and may be of benefit to others (and I acknowledge saving people from having to type millions of parenthesis over their lifetime).

**Comment by DavidArno** [comment]
I'm hurt. How could you doubt a man of my integrity? 😆

**Comment by CyrusNajmabadi** [comment]
> Would be legal? 

Yes

> When separating on multiple lines, the syntax looks OK, but when on the same line, it starts to get all jumbled together.

Then don't write it that way. :)   It's not like parens would be disallowed.  They would just not be required.  Similar to how we don't require curlies.  It's up to you if you want to mandate them for your own code or not depending on situation.

**Comment by casperOne** [comment]
> Then don't write it that way. :) It's not like parens would be disallowed. They would just not be required. Similar to how we don't require curlies. It's up to you if you want to mandate them for your own code or not depending on situation.

Someone didn't read the rest of my response =P

**Comment by lachbaer** [comment]
>  This becomes a stylistic issue though,

Exactly! Like the placing of curly braces or so many other different things. A common style will establish nevertheless, one that is primarily teached in books. But *then* you should also allow `until` when everything becomes so 'verbose'.

**Comment by jnm2** [comment]
```c#
if !foo.Bar() foo.Baz();
```

Meh.

**Comment by jnm2** [comment]
I predict pattern matching complications.

```c#
if obj is Foo foo foo();
```

**Comment by jnm2** [comment]
Knew it wouldn't be long. Ambiguity:
```c#
if obj is Foo foo (x).Bar();
```

`if (obj is Foo) foo (x).Bar();` or `if (obj is Foo foo) (x).Bar();`?

**Comment by lachbaer** [comment]
Well, *very obviously* the use of a *statement-block* must be forced in case the condition parentheses are dropped. Not a big drawback, because most flow statements I see around have braced statement blocks anyway.

Another advantage *to me* is, that it makes one-line `if`-statements, e.g. for early returns, optically better readable.
```cs
if obj is null { return; }
// vs.
if (obj is null) { return; }
```
The second is normally spread over 4 lines in total, I guess that that coding style exists, because two clasped regions in one line is *"ugly"*.

With only one braced block however it plays in the same leage as other *used* constructs.
```cs
public Student(string Name, int Age) : base(string Name) { _age = Age; }
string Name { get; private set; }
if foo is null { bar(); }
```
(First one has other `(...)`, but nevertheless I've seen this construct many times.)

**Comment by CyrusNajmabadi** [comment]
> (First one has other (...), but nevertheless I've seen this construct many times.)

Every ```if`` in C# has always had ```(...)```, so you've certainly "seen [that] construct many times" as well :)

**Comment by lachbaer** [comment]
@CyrusNajmabadi **?**

**Comment by CyrusNajmabadi** [comment]
>Ambiguity:

C# is rife with ambiguities :)   The only rules are:

1. Old code needs to parse as before (unless the benefit so far outweighs that).  For example, we accept that we parse ```Foo(a < b, c > (d) )``` differently that before because the number of people hurt by this is so small, and the new parsing is so worthwhile.
2. With ambiguities, you have to specify the disambiguation rule.  

c# 1.0 shipped with this ambiguity:

```(a)-b```.  But we simply defined how it would be parsed and moved on.  We'd do the same here.  Likely we'd take the maximal approach to sub production parsing.  i.e. since we could, locally, parse out a pattern here, we woudl, leading to the interpretation:

```if (obj is Foo foo) (x).Bar();```

**Comment by CyrusNajmabadi** [comment]
@lachbaer You said: 

> because two clasped regions in one line is "ugly". ... With only one braced block however it plays in the same leage as other used constructs.

But you then referenced a construct that has two clasped regions in one line. You said that that was ok though because you'd seen that construct many times.

But, of course, if it's ok because you've seen it many times, then certainly a multi-clasped 'if' on a single line is ok because you'll have seen ```(...)``` with ```if```s 100% of the time :)

**Comment by lachbaer** [comment]
@CyrusNajmabadi I thought it was very obvious what I wrote. But especially for you:
* `public Student(string Name, int Age) : base(string Name) { _age = Age; }` seen many times.
* `if (obj is null) { return; }` seen rather seldom (spread over 4 lines normally).

**Comment by lachbaer** [comment]
If removing the need for the condition parantheses, what is a *big* change for a B/C-derived language since their existence, then **IMO** the enforcement of a braced statement-block for that new syntax should also be introduced.

**Comment by CyrusNajmabadi** [comment]
> then IMO the enforcement of a braced statement-block for that new syntax should also be introduced.

This seems like one step forward, one step back.  :)

**Comment by lachbaer** [comment]
> This seems like one step forward, one step back. :)

I understand what you mean, but sometimes a step back isn't a bad step if you're at the edge of an abyss 😀

I mean, you once wrote that you primarily focus on features for educated programmers.  Most advanced code that I have seen use statement-blocks always anyhow (we're talking about control statements here). With this change you had the unique possibility to write history and shift it for the better, for the sake of clarity to all developers.

> `if obj is Foo foo (x).Bar();`

Even if you resolved the amiguities by definition, there is the chance that other developers meant it differently. Why introducing a just possible trap when you can avoid it in the first place? Hasn't C# a leading paradigm of avoiding errors before they occur?

**Comment by CyrusNajmabadi** [comment]
that's a fair point.  But i don't think it's our place to mandate that sort of thing at the language level.  i think it would be much more appropriate to just give users style options in the IDE so they can decide on stylistic things themselves.  For example, i have to accept that people that want this, will want to write:

```c#
if (!(o is string))
    return;
```

And will want to convert it to:

```c#
if !(o is string)
    return;
```

If we say "if you want to remove hte parens, now you have to write:

```c#
if !(o is string)
{
    return;
}
```

Then that's going to kill the value of this feature for many users.

**Comment by CyrusNajmabadi** [comment]
Note: i'm not dismissing this concern.  It would definitely be a deep one.  But i would not consider it a killer for this idea.  We would have to consider what we would do here, and it might entail carving out a set of allowed/disallowed things.  I'm just trying to clarify that it's unlikely that we could introduce a code-reduction feature, and then tack on a required code-increase feature :D

**Comment by yaakov-h** [comment]
@lachbaer 

> then IMO the enforcement of a braced statement-block for that new syntax should also be introduced.

Once again, this is a job for a Roslyn Analyzer. You could make the case to include one with Roslyn/VS/`.editorconfig`, but other people can have different opinions and switch off the analyzer rule or not include it in the first place.

I think it would be strange to have `if expr Thing();` become a valid single line of C#, but after playing around with Swift I'm very much in favour of `if expr { block }` as regular syntax.

@CyrusNajmabadi Is there any precedent etc. of creating a compiler warning or error from such an ambiguity as the one defined above, rather than quietly parsing it as one or the other?

**Comment by CyrusNajmabadi** [comment]
> Is there any precedent etc. of creating a compiler warning or error from such an ambiguity as the one defined above, rather than quietly parsing it as one or the other?

Nope.  There's no precedence afaik.  We've always just picked a way to parse.  Generics were the most obvious examples of that.  It added ambiguities to the language, and we went in the direction of picking one of hte directions to go with (which was not the previous interpretation).

**Comment by CyrusNajmabadi** [comment]
So, we could go a couple of directions here:  

1. Change the grammar to be:  ```if     !_opt    (    expr    )     EmbeddedStatement```.  If we do this, there's no ambiguity problem.  In effect, all we're doing is allowing an optional ```!```.
2. Change the grammar to be:  ```if    expr   EmbeddedStatement```, but deal with potential ambiguity fallouts.  One way we might address this would be to simply state that "expr" had restrictions or disambiguiation rules attached to it.  For example, we might restrict a 'named pattern' from being used.  it's one of the few expressions that allow for this ending ```word word``` form that then ends up potentially causing problems like this.  

Right now i have no idea which i prefer :)

**Comment by yaakov-h** [comment]
'named pattern' being `is Foo foo`?

**Comment by CyrusNajmabadi** [comment]
Yes.  You caught me mid-edit :)

**Comment by lachbaer** [comment]
> we could introduce a code-reduction feature

I don't see it as such. I rather see it as a clarity feature. As far as I understood the frequently upcoming request for `ifnot`, `if !(..)` and alike arises from the wish of a better read- and understandability. And removing the parathesis (code-reduction) is just a possibility to achieve this goal.

Incidentially @CyrusNajmabadi as a team member came up with this idea. I guess nobody had ever believed that a possible request for removing a classical C-construct would be considered as a solution 😇 

```cs
if !(o is string)
    return;
```
I believe the benefits of forcing statement-blocks to prevent devious expression constructs outweigh the possible cases where somebody wants to use that style above!

By introducing "clampless-conditions" with statement-blocks-only there will exist a *further* way to express something. One that is syntactically completely new to the C-style. Why introducing another one - with single substatements - when you can already express the same so many different ways? And what are the odds that the above style is allowed by guide lines, but `if !(o is string) \n { return; } ` hasn't the same value?

Also, there are other constructs where braces are mandatory. Leaving expression bodies aside every method must have braces. `try` and `catch` need them for a reason, though they *could* be nested without 'em when just *defining* a hierarchy (e.g. `finally` breaks the inner nest). Llambdas that start with a control statement need the braces `(int i) => { switch (i) { case 1: return 2; default: return 3; }`.

And as stated above, the vast majority of advanced programmers are probably already using braces anyway. The rest can keep using the old constructs or convert. The benefits of clarity outweigh traditions used by a few.

However, I may completely be mistaken by the number of (min. average) developers who are not using braces in their production code! 🖖 

> Once again, this is a job for a Roslyn Analyzer. You could make the case to include one with Roslyn/VS/.editorconfig, but other people can have different opinions and switch off the analyzer rule or not include it in the first place.

Absolutely agree. You already have answered to your comment yourself 😉 Nevertheless my doubts for allowing single blockless statements still last.

**Comment by lachbaer** [comment]
Btw, when talking about uncluttering conditions, has anyone an idea on how `!(o is string)` can be expressed better without introducing `isnot`? Maybe `o is! string`, in light accordance with `!=`?
```cs
if obj is! string { return; }    // to be read as "isn't"
```

**Comment by DavidArno** [comment]
My vote is definitely for "Change the grammar to be: `if expr EmbeddedStatement`, but deal with potential ambiguity fallouts."

In reality, whilst we can all dream up ugly ambiguities, sure most devs want to be able to read their own code and so will avoid writing such ugly code?

The following seems nice and readable to me, so I'd use the optional parentheses feature here:
```cs
if anyTestFailed return;
```
The following less so, so I wouldn't write it that way:
```cs
if
(
!!!!                       anyTestFailed
){


                                              return;
                                                        }
```
I can make ugly code with existing rules and ambiguities already exist. So the fact that folk can make ugly code from this proposal probably shouldn't be a reason to dismiss it.

**Comment by yaakov-h** [comment]
@lachbaer: How about `if obj.IsNotOfType<string>() { return; }`?

```
public static bool IsNotOfType<T>(this T obj) => !(obj is T);
```

**Comment by lachbaer** [comment]
@yaakov-h `!(obj is type)` is a too common expression to have a wrapper method for it, imo.

**Comment by DavidArno** [comment]
> Btw, when talking about uncluttering conditions, has anyone an idea on how `!(o is string)` can be expressed better without introducing isnot?

Again, to my mind:
```cs
if !o is string return;
```
does the job just fine.

**Comment by yaakov-h** [comment]
@lachbaer I haven't seen it in many places, myself.

@DavidArno What if `o` is a `bool`?

**Comment by yaakov-h** [comment]
Actually, I think for negative type checking / negative pattern matching that #138 would work quite well:

```
void Foo(object o)
{
    guard (o is Bar b) else return;

    // Do something with `b`...
}
```

**Comment by lachbaer** [comment]
> if !o is string return;

See specifications chapter 7.3.1 "operator precedence". `is` has less precedence than `!`. This one shouldn't work the way you intend it to.

**Comment by lachbaer** [comment]
@DavidArno I just fear that especially less experienced programmers will write expressions like
```cs
if obj is Foo foo(x).Bar();
```
and produce errors or get confused why their expression does not compile. Nevertheless I get your point that an experienced programmer would care for ambiguities already at coding time, as he does with parentheses when he doesn't know the precedence table by heart.

**Comment by lachbaer** [comment]
@DavidArno 
```cs
if (!value1 is double) return;
//Error CS0023	Operator '!' cannot be applied to operand of type 'double'	
```

**Comment by DavidArno** [comment]
@lachbaer, good point. That'll be why I can write the following today and end with `False` being printed:
```cs
var x = true;
if (!x is bool y) ; 
var z = y;
Console.WriteLine(z);
```

But really, who writes ugly ambiguous code like this in real life ;)

**Comment by lachbaer** [comment]
@DavidArno Alone the fact that you came up with `if !o is string return;` shows the source of potential errors. And you *are* an experienced developer. That's all I'm ever talking about when I state the term *human factors* over and over 😀

**Comment by DavidArno** [comment]
@lachbaer,
Given that the same problem occurs with:
```cs
if (!o is string) return;
```
then no new potential error is being introduced.

**Comment by lachbaer** [comment]
@DavidArno No, just in principle 😃

**Comment by jnm2** [comment]
I don't see why you'd go to the immense trouble of removing parens if another solution which is more widely useful is an `isnot` keyword.

```
if (x isnot Foo) return;
```

**Comment by lachbaer** [comment]
I have [prototyped](https://github.com/lachbaer/roslyn/tree/clamplessif) the Paren'less Condition Statements (and `foreach`) and ran into following problem:
```cs
foreach var item in new List<int>() {
    Console.WriteLine(item);
    // Error	CS0103 The name 'item' does not exist in the current context
}
```
I needed to use the Syntax Visualizer to see that the start of the statement block was acutally a collection initializer expression. Mpf... obvious... actually...But it shows how easily the removal of the parenthesis can lead to unproductive errors. Just wanted to show... 😃 

PS: The prototype has problems with `do` and crashes on those.

**Comment by lachbaer** [comment]
@jnm2 "> is an `isnot` keyword" We wanted to avoid a new keyword, that's why ;-)

**Comment by jnm2** [comment]
I am going to be very sad if we do not get `isnot` as a new keyword. It's too perfect.

**Comment by lachbaer** [comment]
To give you an insight of the optics. I have just written this silly code with my working prototype:
```cs
if obj != null
    foreach var item in new List<object>(100) { }
        do
            switch item
            {
                case 1: break;
                default: return 2;
            }
        while item is double;
```

**Comment by jnm2** [comment]
> ```c#
> foreach var item in new List<object>(100) { }
> ```

o.O

**Comment by lachbaer** [comment]
@jnm2 You forgot the legs:   o.O=   😀

**Comment by bondsbw** [comment]
@lachbaer Because of the `{ }`, the choices become:

1. Interpret `{ }` as an initializer block `foreach (var item in new List<object>(100) { })`
2. Interpret `{ }` as a statement block `foreach (var item in new List<object>(100)) { }`

For your case only the first case would result in code that would compile.  In the second case, `item` would be out of scope in the `do`/`while` loop.

**Comment by bondsbw** [comment]
But I don't know that we would need to change `foreach`.  There's already a better proposal IMHO, #101, and your code would become

```c#
if obj != null
    from item in new List<object>(100) { }
    do
    {
        do
            switch item
            {
                case 1: break;
                default: return 2;
            }
        while item is double;
    }
```

which I believe is unambiguous.

**Comment by lachbaer** [comment]
if even the braces from the `from-do` could be ommited it can be written
```cs
if obj isnot null
from item in new List<object>(100) { obj }
do do switch item
{
    case 1: break;
    default: return 2;
} while item is double;   
```
The so called "duh-duh-switch" construct 😈

**Comment by bondsbw** [comment]
```c#
	if 
(
		obj
!=null )foreach( 
var item in new List
<
	object
>(	
 100
){})do switch 
	(
	item
	)
           {  case
		   
1
:break;default
:return
2;     }while(
item is 
					double)
;
```
Any legal code can suck if you have no desire to adhere to a decent style.

**Comment by jnm2** [comment]
It's not just a matter of legal syntax either. As @CyrusNajmabadi likes to say, 99% of the time the IDE is assisting the user, it's while the syntax is broken or incomplete.

**Comment by whoisj** [comment]
> I needed to use the Syntax Visualizer to see that the start of the statement block was acutally a collection initializer expression. Mpf... obvious... actually...But it shows how easily the removal of the parenthesis can lead to unproductive errors. Just wanted to show

Yes, but these errors are parse errors and easily trapped by the compiler and rectified by the developer during the parse phase. These are not run-time or other insidious bugs.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Given that the removal of the parenthesis immediately introduces syntax ambiguities around initializers I think that should shut down that option.  Having the compiler conditionally require the parenthesis in those cases to disambiguate would only make the syntax messy and it immediately complicates any attempt to introduce initializer-like syntax elsewhere in the language.  I could see it further complicating the parsing of property patterns as they are currently proposed.

That's in addition to the various syntax examples explored in this thread.  I don't find them more readable.  I'd suspect that the only people who do are those who come from a background of working with languages that either don't have or don't require said parenthesis already.

**Comment by lachbaer** [comment]
The cases that you need collection or object initializers in a condition is rather rare, I strongly believe. For `foreach` they might however occur. I only removed the paran's on `foreach` as well, because I could 😀 

For `switch` on the other side, though that part isn't a condition either, it makes sense again. Especially when `switch` becomes an expression syntax additionally.

I guess, that this whole discussion mainly concers conditions with `isnot` and `until`. Boolean expression and comparisons can be reformed quite often to prevent `if (!(...)`. Negative method invocations on the other hand are quite common and should already catch the readers eye, `if (!HasValue())`.

**Comment by HaloFour** [comment]
Rather rare, but definitely legal.  And if the parenthesis were to be removed from `if` it would be weird to not also remove them from everywhere else that parenthesis are currently required.  I've definitely used initializers in `foreach` and `using` statements.  I could probably find a few cases in `if` statements as well.  Either way, for such a minor syntax change with such a minor benefit the fact that it introduces any amount of ambiguity, to me, makes it not worthwhile.

Not only would a developer have to know that there are two forms of the syntax and be able to switch between reading the both of them, but they would have to learn of the rules when they would be forced to use one form in order to disambiguate themselves.

**Comment by gulshan** [comment]
I don't support removal of parenthesis for `if` and other constructs because that will block my if-else expression proposal at #156 😄 . And I support the original proposal of `if!` and `while!` here. I don't want to have a new `not` keyword or very specific `is not`/`is!` construct.

**Comment by jnm2** [comment]
I'm pretty sure this is the new `private protected`. :smile:

**Comment by Duranom** [comment]
Would prefer the `isnot` or `not` syntax in all honousty as it makes it makes codes more readable which solves many bugs/issues because someone oversaw a `!` or even forget it and overlooked it because the statement was long.

I understand that you don't want a new keywords for every little thing but making the language more readable/understandable is also of big importance, if it means I have to change some variables named because of a readable syntax than I would certainly do that witout hesitation (and would even ask why on earth it is named like isnot or not).
Throwing everywhere '!' for some negation really makes many (newer) features, and possible upcoming ones, harder to read and I would prefer to keep it proper accessible and have it more accessible. (It should be readable by others besides developers)

**Comment by lachbaer** [comment]
I believe that `isnot` is worthwhile the implementation, because it seems to be a commonly used construct and `is ⇔ isnot` goes hand in hand with `== ⇔ !=` and `op_true ⇔ op_false`. I've started [prototyping](https://github.com/lachbaer/roslyn/tree/isnotkeyword) it and quite some work is already done.

**Comment by gafter** [comment]
The parens around an `if` statement's condition are needed to prevent the language from being ambiguous. Here is one way of surfacing the ambiguity:

``` c#
    if condition
        (M)(e1)(e2);
```

``` c#
    if condition(M)
        (e1)(e2);
```

**Comment by alrz** [comment]
Swift/Rust always require block but I think that'd be confusing in C# to be conditional.

**Comment by jordao76** [comment]
Why not have the negation of if/while conditions with the new keywords `unless` and `until`?

```cs
unless (foo is Bar bar)
{
    // stuff
}

until (finished)
{
    // stuff
}

do
{
    // stuff
}
until (finished);
```

**Comment by Joe4evr** [comment]
Because that requires introducing new contextual keywords, which has a higher cost associated. Plus, this version already has a Proposal Champion: #882.

**Comment by alexzzzz** [comment]
I personally never use `!` in conditions due to its poor visibility and incorrect semantics. The only use case where `!` semantically fits is something like this:
```csharp
void Toggle()
{
    isEnabled = !isEnabled;
}
```

Most of the time we don't care what negation of an expression is, we care about what the expression itself is, true or false. 
```csharp
if (!expression) ...
```
This code reads as "if negation of the expression is true then ..."

```csharp
if (!(x is Foo)) ...
```
"If negation of `x is Foo` is true then …", but how cares what the negation of `x is Foo` is? 

A solution that doesn't have these problems (poor visibility and incorrect semantics) and works right now is to use `== false` instead. It's crystal clear, visible, semantically correct and sometimes requires fewer braces.

```csharp
if (expression == false) do_stuff();
```
"If the expression is false then do stuff".

```csharp
do {
    ...
} while (expression == false);
```
"Do stuff while the expression is false"

```csharp
if (x is Foo == false) ...
```
"If `x is Foo` is false then …"

So, I downvoted the proposal because don't see how it is any better that just `== false`.

**Comment by bondsbw** [comment]
I actually tend to read `!` as "not" instead of "the negation of".

So `!(x is Foo)` reads to me as "not x is Foo".  That still sounds a bit awkward, but so does "x is Foo is false".

What sounds natural is "x is not Foo".  That is why I support the Python-based version `x is not Foo` or `x isnot Foo`, or even the more C#-ish `x !is Foo`.

**Comment by mikedn** [comment]
> if !(foo is Bar bar)

Cool kids use `if (foo is Bar is false)` :grin:

**Comment by alrz** [comment]
Ah, just one box away to be actually cool. (https://github.com/dotnet/roslyn/issues/19153)

**Comment by mikedn** [comment]
Presumably the pattern matching code was written by cats, that's why it has so many boxes in it.

**Comment by DavidArno** [comment]
@alexzzzz,
> This code reads as "if negation of the expression is true then ..."

You are literally the first person I've come across that reads `!` as "negation of". It's "not", eg `if (!expression)` literally reads as "if not expression". Out of curiosity, do you read `a || b` as "a in disjunction with b"?

`x == false` is not a good way of expressing a boolean expression (it uses a comparison, rather than logic) and it can catch out the unwary as it is not always equivalent to `!x` (eg, as per [@gafter's comment here](https://github.com/dotnet/csharplang/issues/882#issuecomment-357772904)).

**Comment by gafter** [comment]
Also the first time I’ve heard the suggestion to add “is true” when reading code. Would `if (e is true) ...` be read “if e is true is true”?

**Comment by alexzzzz** [comment]
@DavidArno 
> You are literally the first person I've come across that reads ! as "negation of". It's "not", eg if (!expression) literally reads as "if not expression". Out of curiosity, do you read a || b as "a in disjunction with b"?

1. English is not my native language. I my language the word for the term "negation", I believe, is a much more common word.

2. _"If not enabled or not visible"_ sounds fine to me. _"If not enabled and visible"_ is already ambiguous but sounds still more or less ok. However when I see `if (!(some complex expression with a bunch of braces inside))` I would say _"if the negation of all that thing is true"_ or better _"if the whole thing is false"_, and the latter is actually what we usually mean writing that kind of code. That's why I prefer to write `if (something == false)`.

BTW, I believe all the human languages have their own word for "if", but not all of them have the words for "until" and "unless". Mine doesn't. In my native language  _"until A is B"_ translates roughly to something like _"while A would not have become B"_ (if it makes any sense to you) and _"unless A is B"_ translates to _"only if A is not B"_.

So when I see `until` or `unless` keywords in other computer languages, it requires a bit of brain-twisting to realize what's going on. I don't like these keywords, because they are a potential source of misinterpretations.

**Comment by jordao76** [comment]
@alexzzzz, sorry to disagree, but I believe that `until` and `unless` are _exactly_ the keywords that we need to make the language better in this regard, and I don't see how they could be misinterpreted. Unless of course, as you mentioned, that English is not your first language, and you're not used to these words. It's a fact though that all mainstream languages are in English, and I'm sure you, or any other non-native English speaker, like me as well, would easily get used to those words. If we try to impose our transliterated non-English expressions to the language, I think we'd get a very awkward language as a result.

**Comment by HaloFour** [comment]
I'm not sure what `unless`/`until` really brings over `if !`/`while !`.  It's more verbose, but I guess it stands out more.  But the big problem is that it would have to be contextual in order to not break existing code, which could get weird if an identifier of either name already exists.  I know that I've used both `until` and `unless` as identifier names before.

Also, those keywords also don't solve the problem of trying to negate an `is` expression as part of a larger Boolean expression without having to break out the parenthesis.

**Comment by ufcpp** [comment]
It is unfriendly for non English speakers to increase vocabulary `unless`/`until`.

**Comment by jnm2** [comment]
How is that more unfriendly than `void` or `delegate` (noun), which are words I almost never hear used in English conversation? We just sort of accept what `void` means in C# without drawing upon any intuition.

**Comment by ufcpp** [comment]
`void` and `delegate` are rather easy to remember as technical terms, not as natural English words. Some Japanese developers don't know how "void" means in English but do know it means "no return value".
Difficulty of `unless`/`until` is that different letters (`if` and `unless`) are related to each other. As a negated word of `if`, `if not` (`if !`) is much easier than `unless`.

**Comment by jordao76** [comment]
@HaloFour 

> Also, those keywords also don't solve the problem of trying to negate an is expression as part of a larger Boolean expression without having to break out the parenthesis.

I usually solve that problem with some [refactoring](https://refactoring.com/catalog/extractVariable.html).

**Comment by HaloFour** [comment]
@jordao76

> I usually solve that problem with some refactoring.

While I largely agree, that is substantially more verbose and does not work well when it comes to analysis for definite assignment and pattern matching.  For `while` and `do` loops that would also lead to leaky scope which has further ramifications on closures (which is why leaky scope was backpedaled in those cases.)

**Comment by jordao76** [comment]
@ufcpp 

> It is unfriendly for non English speakers to increase vocabulary unless/until.

It is unfriendly for non-English speakers to have an English-based language. But, as programmers, we get used to it. I'm sure you can deal with extra keywords, as might have been the case with, e.g., `async` and `await`.

@HaloFour, thanks for explaining that.

**Comment by alrz** [comment]
>It is unfriendly for non-English speakers to have an English-based language

Absolutely not. I cannot imagine programming in Persian. that would probably read like Shahnameh by Ferdowsi.

**Comment by jordao76** [comment]
@alrz, I meant that as a very general comment. If you don't know _any_ English, that would be unfriendly at first, but then you'd certainly get used to it. I agree with you, of course, I can't imagine working with a non-English based language.

**Comment by yaakov-h** [comment]
I find that in spoken language, `until` usually comes after the action. For example, "wait here until I get back" is more natural than "until I get back, wait here".

This would make it more suitable for a `do`...`while` sort of control structure, where the condition goes at the end.

Unfortunately, this makes for rather horrible readability IMO - for example, Ruby has statements such as:

```ruby
puts "This string will not be printed" unless true
```

... which trips me up every time.

**Comment by ufcpp** [comment]
@jordao76 
Yes, `async`/`await` pair is difficult. Some one misunderstands `await` as "asynchronouslv wait". It's difficult but there is no word more appropriate than `await`. On the other hand, `until` can be replaced with simpler and already existing words `if!`.

**Comment by alexzzzz** [comment]
@gafter 
> Also the first time I’ve heard the suggestion to add “is true” when reading code. Would if (e is true) ... be read “if e is true is true”?

```csharp
if (control.Validate()) { } // WinForms API
if (int.TryParse(s, out var n)) { }
```
If TryParse… what? If it returns true. If its result is true.

```csharp
if (!(enabled && visible)) { }
```
"If not enabled and visible" is not technically correct, "if not enabled or not visible" is not what is written. Correct reading is "If enabled and visible are not both true". Maybe native speakers can come up with a better option without _be true_/_not be true_/_be not true_, but I can't.

**Comment by jnm2** [comment]
Or "if not both enabled and visible."

**Comment by bondsbw** [comment]
I don't try to convert to unparenthesized English.

**Comment by DavidArno** [comment]
You can sort of express such logical expressions in English:
```
if !(a && b && c) // "if not a and b and c"
if (!a && b && c) // if not a, and b and c"
```
With the "," being vocalised as a slight pause. But such an approach quickly breaks down for non-trivial examples and is highly error prone.

**Comment by Joe4evr** [comment]
> `if (e is true)` is just redundant when e is a bool.

But it's not when `e` is a `bool?`. And that's relatively easy to get these days by having a `?.` in the expression.

**Comment by orthoxerox** [comment]
@DavidArno That's a spoken equivalent of a whitespace-sensitive language!

**Comment by DavidArno** [comment]
@orthoxerox,

Ah, English, don't you just love it? 🤣

**Comment by weitzhandler** [comment]
I prefer an `isnot` operator instead, just as a simple syntactic sugar.

**Comment by cryolithic** [comment]
I personally quite like the idea of negation of the statement. I tend to work with a lot of juniors, and something like this would make many simple boolean errors go away.

**Comment by Freakhealer** [comment]
> There is no actual syntactic need/ambiguity it addresses. 
> Another way to think about this is the grammar was always `if expr`, but there was a normative rule that stated that `expr` needed to be a parenthesized-expression.

Just wanna mention that if you remove parenthesis then you need to specify when the expression ends somehow. Either make block compulsory `{ }`, or like, python you use `:` since in c# you can do:

`if isLogged && isActive GiveAccess();` In my opinion gets a bit confusing 

You could `if isLogged && isActive {GiveAccess();}` but then why not keep `if (isLogged && isActive) GiveAccess();`

**Comment by jkalamarz** [comment]
I'd also love if negation could be move to the very last element of the nested members notation:
```if (isDone && !foo.VeryLongMethodName().MyElement.GetList().Any())```
could be written
```if (isDone && foo.VeryLongMethodName().MyElement.GetList().!Any())```

**Comment by Freakhealer** [comment]
I see where you are getting at but doesn't look great imo
if it matters "very long variable name" appearance you can assign it to a variable:
`var anyElement = foo.VeryLongMethodName().MyElement.GetList().Any()`
then will look better than either of the above options (imo)
`if (isDone && !anyElement)`
I personally never had any issue negating on long nested member, and usually go for the simplest and most direct syntax if I will only need the value once. but anyway maybe this is not the thread to discuss this topic since its different suggestion altogether.

**Comment by Thaina** [comment]
Or `if (isDone && foo.VeryLongMethodName().MyElement.GetList().NotAny())` (another extension method)

**Comment by bernd5** [comment]
Or a `bool` extension method like:

```csharp
if (isDone && foo.VeryLongMethodName().MyElement.GetList().Any().Not())
{
  //do stuff
}
```

**Comment by Thaina** [comment]
Actually I think `NotAll` and `NotAny` is common enough to included in LINQ

And it's not always give the exact same result with negation. `!Any((b) => b)` is not behave the same as `All((b) => !b)` such as when it was empty
